### DISCUSSION TITLE
Train Loader and Test Loader

### DISCUSSION THREAD

**Discussion by riskiem** [pyg-team/pytorch_geometric#2861] [Q&A]
I have a graph as follows:

```python
f1 = np.random.randint(10, size=(21))
f2 = np.random.randint(10,20, size=(21))
f3 = np.random.randint(20,30, size=(21))
f_final = np.stack((f1,f2,f3), axis=1)
capital = 2*f1 + f2 - f3
f1_t = torch.from_numpy(f1)
f2_t = torch.from_numpy(f2)
f3_t =  torch.from_numpy(f3)
capital_t =  torch.from_numpy(capital)
capital_t = capital_t.type(torch.FloatTensor)
x = torch.from_numpy(f_final)
x = x.type(torch.FloatTensor)
edge_index = edge_index = torch.tensor([[0,1],[0,5],[0,4],[1, 2],
                           [1,3],[1,4],[2,7],[2,6],[2,3],[3,6],[3,5],[4,5],
                           [5,9],[6,10],[6,8],[6,7],[6,11],[7,12],[3,8],[8,9],
                           [8,10],[8,14],[9,13],[9,10],[10,11],[12,4],[12,14],
                           [12,13],[9,13],[13,5],[14,4],[1,15],[4,15],
                           [3,15],[2,16],[5,16],[13,16],[17,9],[17,12],
                           [14,17],[2,18],[7,18],[11,18],[19,11],[19,12],
                           [19,10],[20,4],[20,5],[13,20]], dtype=torch.long)
#edge_attr = torch.tensor([[5],[-5],[2],[-2],[3],[-3],[6],[-6],[3],[-3],[1],[-1],[4],[-4],[2],[-2],[7],
       ##                   [-7],[11],[-11],[12],[-12],[4],[-4],[3],[-3],[2],[-2],[5],[-5],[1],[-1],[2],[-2],[3],[-3],], dtype=torch.float)
edge_attr = torch.tensor([[1],[4],[3],[5],[2],[3],[7],[1],[3],[6],[1],[3],[2],[4],[2],[7],[6],[5],
                          [11],[7],[3],[7],[3],[4],[12],[2],[4],[5],[3],[1],[2],[3],[4],[7],[6],[15],[3],[5],[4],
                          [4],[6],[3],[5],[7],[7],[6],[9],[1],[7]],dtype=torch.float)
data = Data(x = x, edge_index = edge_index.t().contiguous(), y = capital_t, edge_attr=edge_attr )
print(data)
```

I am doing a test/train split using the following:
```python
#Split into test and train
data.train_mask = torch.zeros(data.num_nodes, dtype=torch.uint8)
data.train_mask[:int(0.8 * data.num_nodes)] = 1 #train only on the 80% nodes
data.test_mask = torch.zeros(data.num_nodes, dtype=torch.uint8) #test on 20 % nodes 
data.test_mask[- int(0.2 * data.num_nodes):] = 1
```

However, for more generic operationalization, I feel creating train and test loader would be better. Anyone can help me do this for the same?

**Comment by rusty1s** [comment]
You can also copy `data` and add the respective masks to each `*_data` object:
```python
train_data = copy.copy(data)
train_data.mask = ...
test_data = copy.copy(data)
test_data.mask = ...
```

**Reply by riskiem** [reply]
I see. What I am looking to do is calculate the degrees in the PNAConv. 

`deg = torch.zeros(5, dtype=torch.long)
for data in train_dataset:
    d = degree(data.edge_index[1], num_nodes=data.num_nodes, dtype=torch.long)
    deg += torch.bincount(d, minlength=deg.numel())`

So, I was looking to put my data into a format that can do the above.

**Reply by rusty1s** [reply]
When learning on single graphs, you can simple use the bincount of degrees of this single graph.
